src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSearchValues.cs

```diff
@@ -0,0 +1,314 @@
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            DiagnosticId,
+            CreateLocalizableResourceString(nameof(UseSearchValuesTitle)),
+            CreateLocalizableResourceString(nameof(UseSearchValuesMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            description: CreateLocalizableResourceString(nameof(UseSearchValuesDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        protected abstract bool IsConstantByteOrCharArrayVariableDeclaratorSyntax(SemanticModel semanticModel, SyntaxNode syntax, out int length);
+
+        protected abstract bool IsConstantByteOrCharReadOnlySpanPropertyDeclarationSyntax(SemanticModel semanticModel, SyntaxNode syntax, out int length);
+
+        protected abstract bool IsConstantByteOrCharArrayCreationSyntax(SemanticModel semanticModel, SyntaxNode syntax, out int length);
+
+        protected abstract bool ArrayFieldUsesAreLikelyReadOnly(SyntaxNode syntax);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemBuffersSearchValues, out _) ||
+                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlySpan1, out var readOnlySpanType))
+                {
+                    return;
+                }
+
+                var indexOfAnyMethods = GetIndexOfAnyMethods(context.Compilation);
+
+                context.RegisterOperationAction(context => AnalyzeInvocation(context, indexOfAnyMethods, readOnlySpanType), OperationKind.Invocation);
+            });
+        }
+
+        private void AnalyzeInvocation(OperationAnalysisContext context, HashSet<IMethodSymbol> indexOfAnyMethodsToDetect, INamedTypeSymbol readOnlySpanType)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+
+            if (!indexOfAnyMethodsToDetect.Contains(invocation.TargetMethod))
+            {
+                return;
+            }
+
+            Debug.Assert(invocation.Arguments.Length is 1 or 2);
+            IArgumentOperation valuesArgument = invocation.Arguments[^1];
+
+            bool isStringIndexOfAny = invocation.TargetMethod.ContainingType.SpecialType == SpecialType.System_String;
+
+            if (isStringIndexOfAny && invocation.Parent is IConditionalAccessOperation)
+            {
+                // We don't flag uses like "string?.IndexOfAny(char[])"
+                // as it's not trivial to rewrite to the span-based variant.
```

Comment:
> as it's not trivial to rewrite

Is this statement with respect to the code fixer OR it applies even for users to manually update their code? If it is the latter, we can still consider reporting a diagnostic for this case and just bail out upfront from the code fix provider.

Reply:
Both, as they have slightly different semantics. `string.IndexOfAny(values, offset)` will return the index including the `offset`, whereas `string.AsSpan(offset).IndexOfAny(values)` wouldn't.

The code fix would then have to also emit an extra branch like
```c#
int index = string.AsSpan(offset).IndexOfAny(values);
return index <= 0 ? index : index + offset;
```

which may not fit well into user code